ldap/servers/slapd/daemon.c

```diff
@@ -1202,133 +1292,117 @@ clear_signal(struct POLL_STRUCT *fds)
     return 0;
 }
 
-static int first_time_setup_pr_read_pds = 1;
-static int listen_addr_count = 0;
-
-static void
-setup_pr_read_pds(Connection_Table *ct, PRFileDesc **n_tcps, PRFileDesc **s_tcps, PRFileDesc **i_unix, PRIntn *num_to_read)
+static PRIntn
+setup_pr_accept_pds(PRFileDesc **n_tcps, PRFileDesc **s_tcps, PRFileDesc **i_unix,
+    struct POLL_STRUCT **fds)
 {
-    Connection *c = NULL;
-    Connection *next = NULL;
     LBER_SOCKET socketdesc = SLAPD_INVALID_SOCKET;
-    int accept_new_connections;
-    static int last_accept_new_connections = -1;
     PRIntn count = 0;
-    slapdFrontendConfig_t *slapdFrontendConfig = getFrontendConfig();
     size_t n_listeners = 0;
 
-    accept_new_connections = ((ct->size - g_get_current_conn_count()) > slapdFrontendConfig->reservedescriptors);
-    if (!accept_new_connections) {
-        if (last_accept_new_connections) {
-            slapi_log_err(SLAPI_LOG_ERR, "setup_pr_read_pds",
-                          "Not listening for new connections - too many fds open\n");
-            /* reinitialize n_tcps and s_tcps to the pds */
-            first_time_setup_pr_read_pds = 1;
-        }
-    } else {
-        if (!last_accept_new_connections &&
-            last_accept_new_connections != -1) {
-            slapi_log_err(SLAPI_LOG_ERR, "setup_pr_read_pds",
-                          "Listening for new connections again\n");
-            /* reinitialize n_tcps and s_tcps to the pds */
-            first_time_setup_pr_read_pds = 1;
-        }
+    // How many fds do we have?
+    if (n_tcps != NULL) {
+        PRFileDesc **fdesc = NULL;
+        for (fdesc = n_tcps; fdesc && *fdesc; fdesc++, count++) { }
     }
-    last_accept_new_connections = accept_new_connections;
+    if (s_tcps != NULL) {
+        PRFileDesc **fdesc = NULL;
+        for (fdesc = s_tcps; fdesc && *fdesc; fdesc++, count++) { }
+    }
+#if defined(ENABLE_LDAPI)
+    if (i_unix != NULL) {
+        PRFileDesc **fdesc = NULL;
+        for (fdesc = i_unix; fdesc && *fdesc; fdesc++, count++) { }
+    }
+#endif
 
+    struct POLL_STRUCT *myfds = (struct POLL_STRUCT *)slapi_ch_calloc(1, (count + 1) * sizeof(struct POLL_STRUCT));
```

Comment:
A minor point: Could you move myfds declaration to the top of the function (its scope covers the whole function)